File: Blind150/9.Backtrack/Go/permutations.go
// https://leetcode.com/problems/permutations/description/
func permute(nums []int) [][]int {

	ans := make([][]int, 0)

	seen := make(map[int]bool)

	N := len(nums)

	var dfs func(tmp []int, seen map[int]bool)

	dfs = func(tmp []int, seen map[int]bool) {

		if len(tmp) == N {
			subset := make([]int, N)

			copy(subset, tmp)

			ans = append(ans, subset)

			return
		}

		for _, num := range nums {

			if !seen[num] {

				tmp = append(tmp, num)
				seen[num] = true

				dfs(tmp, seen)

				tmp = tmp[:len(tmp)-1]
				delete(seen, num)
			}

		}

	}

	dfs([]int{}, seen)

	return ans

}

File: Blind150/9.Backtrack/Python/permutations.py
# https://leetcode.com/problems/permutations/description/
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:


        seen = set()
        ans = []
        N = len(nums)

        def dfs(tmp, seen):

            if len(tmp) == N:
                ans.append(tmp[::])
                return

            for i in nums:

                if i not in seen:
                    tmp.append(i)
                    seen.add(i)
                    dfs(tmp, seen)
                    tmp.pop()
                    seen.remove(i)

        dfs([], seen)

        return ans
